Add support for dict generator input_fn in learn_io

abhitopia opened the pull request:
This issue if related to #5546, where it is claimed that `Datafeeders` will become deprecated.

As such, this branch has following changes 
- [ ] Added `_GeneratorFeedFn` class
- [ ] `generator_input_fn` in `generator_io.py`
- [ ] support for `generator_input_fn` in `enqueue_data()`
- [ ] added unitest in `generator_io_test.py`
- [ ] refactored code keeping indent spacing to 2.

agistrueai committed:
Branch to test_tensorflow

agistrueai committed:
using types.GeneratorType for validation

agistrueai committed:
changed to receving generator function

This allows to run multiple epochs of a finite generator

agistrueai committed:
using placeholders as keys to feed dictionary

agistrueai committed:
added correct counter for the index_placeholder

added corrections to tests

Commit:
removing extra spaces

tensorflow-jenkins commented:
Can one of the admins verify this patch?

Commit:
doc_string fixes

agistrueai commented:
I consent to contribute to this project with my commits

agistrueai committed:
Changed logic to handle final batch

if the generator ends  it restarts to fill the final batch

agistrueai committed:
Merge branch 'add_dict_iter_input_fn_in_learn_io' of https://github.com/abhitopia/tensorflow into add_dict_iter_input_fn_in_learn_io

abhitopia commented:
Cc @martinwicke

agistrueai committed:
Added check that dictionary is returned from gen

fixes in tests

agistrueai committed:
Fixes in docstrings, tests and target_key bug

agistrueai committed:
All tests fixed and pass

agistrueai committed:
Added generator_input_fn to __init__

abhitopia commented:
@ilblackdragon @martinwicke - this is what you guys suggested in #5546 . please review.

agistrueai committed:
revert repo and branch in dockerfile.devel to original

rmlarsen commented:
@ilblackdragon this is ready for review since @agistrueai  can given his consent to the PR.

abhitopia commented:
@martinwicke , @ilblackdragon , @rmlarsen - it's been a while, can one of the reviewers please take a look. As already mentioned, @agistrueai has already provided his consent.

martinwicke commented on tensorflow/contrib/learn/python/learn/learn_io/generator_io.py:
```diff
@@ -0,0 +1,117 @@
+    KeyError:  `target_key` not a key in next(`x()`)
+  """
+  
+  def input_fn():
```
Maybe call this _generator_input_fn? This name sometimes leaks into strack traces, and should be informative.

agistrueai replied:
Done

martinwicke commented on tensorflow/contrib/learn/python/learn/learn_io/generator_io.py:
```diff
@@ -0,0 +1,117 @@
+  
+  def input_fn():
+    """generator input function."""
+    if not isinstance(x, FunctionType):
```
Why does this need to be a function which when called returns a generator? Why not just take any object with __next__ (any generator-like)?

outcastofmusic replied:
replied above

martinwicke commented on tensorflow/contrib/learn/python/learn/learn_io/generator_io.py:
```diff
@@ -0,0 +1,117 @@
+
+    input_keys = next(x()).keys()
+    if target_key is not None and target_key not in input_keys:
+      raise KeyError('target_key must be present in the yielded dictionary')
```
I think it would be a much better interface to make the generator return two dicts.

outcastofmusic replied:
I agree. I kept it this ways as both pandas_io, and numpy_io, return a dict of tensors for the features and an optional single tensor for the labels, and followed the same pattern. Not sure if I should change it or not

agistrueai replied:
I followed the advice in the end. In the latest commit. the output is either a single dict of tensors or two dicts of tensors. One for the features and one for the labels.

martinwicke commented on tensorflow/contrib/learn/python/learn/learn_io/generator_io.py:
```diff
@@ -0,0 +1,117 @@
+      raise TypeError('x() must be generator ; got {}'.format(type(generator).__name__))
+    data = next(generator)
+    if not isinstance(data, OrderedDict):
+      raise TypeError('x() must yield OrderedDict ; got {}'.format(type(data).__name__))
```
Why require OrderedDict? Why is a dict not enough?

outcastofmusic replied:
It doesn't have to be. It was done because the `feeding_functions.enqueue_data` uses ordered dicts for the numpy_io and I kept it the same for consistency.

agistrueai replied:
Changed it to use a normal dict as suggested.

martinwicke commented on tensorflow/contrib/learn/python/learn/learn_io/generator_io.py:
```diff
@@ -0,0 +1,117 @@
+    
+    features = (queue.dequeue_many(batch_size) if num_epochs is None
+                else queue.dequeue_up_to(batch_size))
+    
```
Single blank line only

agistrueai replied:
Done.

martinwicke commented on tensorflow/contrib/learn/python/learn/learn_io/generator_io.py:
```diff
@@ -0,0 +1,117 @@
+    
+
+    # Remove the first `Tensor` in `features`, which is the row number.
+    if len(features) > 0:
```
Simply `if features:`?

outcastofmusic replied:
because we are not checking if the features exists, but to remove the first item if it has more than one item ( this is done because the feeding functions in  `feeding_functions.enqueue_data` use the first item in the dict to add an index, which gets removed. I am not sure onwhy it is done this way but I followed the same pattern.

agistrueai replied:
Actually removed the unused index key from the returned dict so completely removed this

martinwicke reviewed (changes requested):
I think we should make the generator return the same type of things (tuple of dicts) we want from the function. Otherwise I fear it'll be confusing.

vrv commented:
Friendly ping for @abhitopia

outcastofmusic commented on tensorflow/contrib/learn/python/learn/learn_io/generator_io.py:
```diff
@@ -0,0 +1,117 @@
+    Function, that has signature of ()->(dict of `features`, `target`)
+
+  Raises:
+    TypeError: `x` is not `FunctionType`.
```
It is very important to pass a function instead of a generator. This way the `feeding_functions.enqueue_data` can handle multiple epochs. Each time the generator gets consumed one epoch has passed. And it can re-create it. This way the estimator can handle the number of epochs, just like the other *_io methods

outcastofmusic commented on tensorflow/contrib/learn/python/learn/learn_io/generator_io.py:
```diff
@@ -0,0 +1,117 @@
+    num_threads: Integer, number of threads used for reading and enqueueing.
+
+  Returns:
+    Function, that has signature of ()->(dict of `features`, `target`)
```
Again followed the same pattern as numpy_io and pandas_io

agistrueai committed:
docstring changes based on reviewer comments

outcastofmusic commented on tensorflow/contrib/learn/python/learn/learn_io/generator_io.py:
```diff
@@ -0,0 +1,117 @@
+                       queue_capacity=1000,
+                       num_threads=1):
+  """Returns input function that would feed a generator that yields dictionary
+  of numpy arrays into the model. It is assumed that every dict yielded from
```
Done.

martinwicke replied:
first line is not <80chars though.

outcastofmusic commented on tensorflow/contrib/learn/python/learn/learn_io/generator_io.py:
```diff
@@ -0,0 +1,117 @@
+    def generator():
+      for index in range(10):
+        yield collections.OrderedDict({height: np.random.randint(32,36), 'age':np.random.randint(18,80),
+              "label":np.ones(1)})
```
Done

agistrueai replied:
If I understand the comment correctly, the label is not a magic key, I just used it for example purposes. The generaror_io is going to select for the label dict, the key or list of keys that got passed to the target_key argument

outcastofmusic commented on tensorflow/contrib/learn/python/learn/learn_io/generator_io.py:
````diff
@@ -0,0 +1,117 @@
+    ```
+
+  Args:
+    x: Generator Function, returns a generator  that will yield the data
````
Done.

agistrueai committed:
target returns now a dict of labels

target_key can now be a list of keys to be used to populate the target

agistrueai committed:
added more test functions for the new label dict possible errors

agistrueai commented:
Apologies, I used my other github account @outcastofmusic  to reply to the comments

agistrueai commented on tensorflow/contrib/learn/python/learn/learn_io/generator_io.py:
```diff
@@ -0,0 +1,117 @@
+  of numpy arrays into the model. It is assumed that every dict yielded from
+  the dictionary represents a single sample for every feature
+
+  This returns a function outputting `features` and `target` based on the dict
```
I changed it so that the output is currently a feature dict, or  a feature dict and a target dict, if the target_key is used. I could cover the case where the target_key is a single str. In this case instead of a dict it returns a single tensor. This potentially makes it confusing as to what to expect from the function though as an output

abhitopia commented:
thanks @vrv

agistrueai committed:
if label is a single feature return a tensor (to make it work with estimator object)

moved asserts to begining of generator_input_fn
added test for dictionary with single input as generator return

dandelionmane commented:
@martinwicke PTAL

dandelionmane commented:
@abhitopia Please resolve the conflict

martinwicke commented on tensorflow/contrib/learn/python/learn/dataframe/queues/feeding_functions.py:
```diff
@@ -118,8 +167,8 @@ def __call__(self):
     integer_indexes = [
-        j % self._max for j in range(self._trav, self._trav + self._batch_size)
-    ]
+      j % self._max for j in range(self._trav, self._trav + self._batch_size)
```
4 spaces here (I know, I know, it's weird)

martinwicke commented on tensorflow/contrib/learn/python/learn/dataframe/queues/feeding_functions.py:
```diff
@@ -118,8 +167,8 @@ def __call__(self):
-        j % self._max for j in range(self._trav, self._trav + self._batch_size)
-    ]
+      j % self._max for j in range(self._trav, self._trav + self._batch_size)
+      ]
```
And no extra spaces here

martinwicke commented on tensorflow/contrib/learn/python/learn/dataframe/queues/feeding_functions.py:
```diff
@@ -129,9 +178,9 @@ def __call__(self):
-    ]
+      column[integer_indexes]
+      for column in self._ordered_dict_of_arrays.values()
+      ]
```
Same here: 4 spaces inside, no spaces. Like so:

```
cols = [
    stuffs...
]
```

martinwicke commented on tensorflow/contrib/learn/python/learn/dataframe/queues/feeding_functions.py:
```diff
@@ -166,8 +215,8 @@ def __call__(self):
-        j % self._max for j in range(self._trav, self._trav + self._batch_size)
-    ]
+      j % self._max for j in range(self._trav, self._trav + self._batch_size)
+      ]
```
Same here

martinwicke commented on tensorflow/contrib/learn/python/learn/dataframe/queues/feeding_functions.py:
```diff
@@ -148,7 +197,7 @@ def __init__(self,
     if len(placeholders) != len(dataframe.columns) + 1:
       raise ValueError("Expected {} placeholders; got {}.".format(
-          len(dataframe.columns), len(placeholders)))
+        len(dataframe.columns), len(placeholders)))
```
4 space for continuation after opening param

martinwicke commented on tensorflow/contrib/learn/python/learn/dataframe/queues/feeding_functions.py:
```diff
@@ -80,11 +83,57 @@ def __call__(self):
+    first_sample = next(generator())
+    if len(placeholders) != len(first_sample) + 1:
+      raise ValueError("Expected {} placeholders; got {}.".format(
+        len(first_sample), len(placeholders)))
```
4 space continuation

martinwicke commented on tensorflow/contrib/learn/python/learn/dataframe/queues/feeding_functions.py:
```diff
@@ -53,7 +56,7 @@ def __init__(self,
     if len(placeholders) != 2:
       raise ValueError("_array_feed_fn expects 2 placeholders; got {}.".format(
-          len(placeholders)))
+        len(placeholders)))
```
4 space continuation

martinwicke commented on tensorflow/contrib/learn/python/learn/dataframe/queues/feeding_functions.py:
```diff
@@ -70,8 +73,8 @@ def __call__(self):
-        j % self._max for j in range(self._trav, self._trav + self._batch_size)
-    ]
+      j % self._max for j in range(self._trav, self._trav + self._batch_size)
+      ]
```
4 space continuation, no space at closing brace

martinwicke commented on tensorflow/contrib/learn/python/learn/dataframe/queues/feeding_functions.py:
```diff
@@ -80,11 +83,57 @@ def __call__(self):
-        self._placeholders[0]: integer_indexes,
-        self._placeholders[1]: self._array[integer_indexes]
+      self._placeholders[0]: integer_indexes,
+      self._placeholders[1]: self._array[integer_indexes]
```
4 space continuation

martinwicke commented on tensorflow/contrib/learn/python/learn/dataframe/queues/feeding_functions.py:
```diff
@@ -80,11 +83,57 @@ def __call__(self):
+      raise errors.OutOfRangeError(None, None,
+                                   "Already emitted %s epochs." % self._epoch)
+
+    list_dict = collections.OrderedDict({self._index_placeholder: np.arange(self._index, self._index + self._batch_size)})
```
Can you make sure lines are broken at 80 characters?

martinwicke commented on tensorflow/contrib/learn/python/learn/dataframe/queues/feeding_functions.py:
```diff
@@ -97,7 +146,7 @@ def __init__(self,
     if len(placeholders) != len(ordered_dict_of_arrays) + 1:
       raise ValueError("Expected {} placeholders; got {}.".format(
-          len(ordered_dict_of_arrays), len(placeholders)))
+        len(ordered_dict_of_arrays), len(placeholders)))
```
4 space continuation

martinwicke commented on tensorflow/contrib/learn/python/learn/dataframe/queues/feeding_functions.py:
```diff
@@ -195,16 +244,17 @@ def enqueue_data(data,
                  enqueue_size=1,
                  num_epochs=None):
-  """Creates a queue filled from a numpy array or pandas `DataFrame`.
+  """Creates a queue filled from a numpy array or pandas `DataFrame` or a `Generator` yield dicts of numpy arrays.
```
long line, 80 chars max please

martinwicke commented on tensorflow/contrib/learn/python/learn/dataframe/queues/feeding_functions.py:
```diff
@@ -231,57 +281,64 @@ def enqueue_data(data,
-          dtypes.as_dtype(col.dtype) for col in data.values()
-      ]
+        dtypes.as_dtype(col.dtype) for col in data.values()
+        ]
```
4 space continuation

martinwicke commented on tensorflow/contrib/learn/python/learn/dataframe/queues/feeding_functions.py:
```diff
@@ -231,57 +281,64 @@ def enqueue_data(data,
       types = [
-          dtypes.as_dtype(dt) for dt in [data.index.dtype] + list(data.dtypes)
-      ]
+        dtypes.as_dtype(dt) for dt in [data.index.dtype] + list(data.dtypes)
```
here too

martinwicke commented on tensorflow/contrib/learn/python/learn/dataframe/queues/feeding_functions.py:
```diff
@@ -231,57 +281,64 @@ def enqueue_data(data,
-          "num_epochs is applied per thread, so this will produce more "
-          "epochs than you probably intend. "
-          "If you want to limit epochs, use one thread.")
+        "enqueue_data was called with num_epochs and num_threads > 1. "
```
revert this chunk, 4 space was correct

martinwicke commented on tensorflow/contrib/learn/python/learn/dataframe/queues/feeding_functions.py:
```diff
@@ -231,57 +281,64 @@ def enqueue_data(data,
-          "num_epochs. This will create multiple threads, all reading the "
-          "array/dataframe in order adding to the same shuffling queue; the "
-          "results will likely not be sufficiently shuffled.")
+        "enqueue_data was called with shuffle=True, num_threads > 1, and "
```
4 space was correct

martinwicke commented on tensorflow/contrib/learn/python/learn/dataframe/queues/feeding_functions.py:
```diff
@@ -231,57 +281,64 @@ def enqueue_data(data,
-          "This will create multiple threads, all reading the "
-          "array/dataframe in order. If you want examples read in order, use"
-          " one thread; if you want multiple threads, enable shuffling.")
+        "enqueue_data was called with shuffle=False and num_threads > 1. "
```
4 space

martinwicke commented on tensorflow/contrib/learn/python/learn/dataframe/queues/feeding_functions.py:
```diff
@@ -231,57 +281,64 @@ def enqueue_data(data,
-          dtypes=types,
-          shapes=queue_shapes,
-          seed=seed)
+        capacity,
```
4 space

martinwicke commented on tensorflow/contrib/learn/python/learn/dataframe/queues/feeding_functions.py:
```diff
@@ -231,57 +281,64 @@ def enqueue_data(data,
       min_after_dequeue = 0  # just for the summary text
       queue = data_flow_ops.FIFOQueue(
-          capacity, dtypes=types, shapes=queue_shapes)
+        capacity, dtypes=types, shapes=queue_shapes)
```
4 space

martinwicke commented on tensorflow/contrib/learn/python/learn/dataframe/queues/feeding_functions.py:
```diff
@@ -294,21 +351,21 @@ def enqueue_data(data,
-              random_start=shuffle,
-              seed=seed_i,
-              num_epochs=num_epochs))
+        get_feed_fn(
```
4 space after (

martinwicke commented on tensorflow/contrib/learn/python/learn/dataframe/queues/feeding_functions.py:
```diff
@@ -294,21 +351,21 @@ def enqueue_data(data,
 
     runner = fqr.FeedingQueueRunner(
-        queue=queue, enqueue_ops=enqueue_ops, feed_fns=feed_fns)
+      queue=queue, enqueue_ops=enqueue_ops, feed_fns=feed_fns)
```
4 space

martinwicke commented on tensorflow/contrib/learn/python/learn/dataframe/queues/feeding_functions.py:
```diff
@@ -294,21 +351,21 @@ def enqueue_data(data,
     full = (math_ops.cast(
-        math_ops.maximum(0, queue.size() - min_after_dequeue),
-        dtypes.float32) * (1. / (capacity - min_after_dequeue)))
+      math_ops.maximum(0, queue.size() - min_after_dequeue),
```
4 space

martinwicke reviewed (changes requested):
I think you took a 2 space thing a little too far.

agistrueai committed:
Merge branch 'master' of https://github.com/tensorflow/tensorflow into add_dict_iter_input_fn_in_learn_io

# Conflicts:
#	tensorflow/contrib/learn/python/learn/dataframe/queues/feeding_functions.py

agistrueai committed:
Added sorting to keys to ensure order

agistrueai commented:
@martinwicke  fixed conflicts with master and made relevant indentation changes. Generator_io is still in learn_io  in the contrib section while the other input functions have been moved to estimator.inputs is this ok?

martinwicke commented:
All of the code should go into contrib, including the private backing class. Can you move it? I'll add comments.

agistrueai commented:
@martinwicke I can move the the _GeneratorFeedFn inside tensorflow.contrib.learn.python.learn.dataframe.queues.feeding_functions.py, but what do I do about the changes in  _enqueue_data (which is located inside tensorflow.python.estimator.inputs.queues.feeding_functions.py

martinwicke commented:
I think the change to _enqueue_data is fine where it is​, it's generally
useful.

agistrueai commented:
@martinwicke  ok I 'll need to ad an import for the _GeneratorFeedFn though

agistrueai committed:
moved _GeneratorFeedFn to contrib

agistrueai committed:
fixed numpy import

martinwicke commented:
Argh. Darn, you're right. Having the object over with _enqueue was better. I'm sorry about that. Can you revert that change?

agistrueai committed:
Revert "Added sorting to keys to ensure order"

This reverts commit 4ad3a0ddee7f263c650fefe9139a95099d457467.

# Conflicts:
#	tensorflow/python/estimator/inputs/queues/feeding_functions.py

agistrueai commented:
reverted :)

agistrueai committed:
removed duplicate generator feed function from contrib feeding_functions.py

agistrueai committed:
indendation fixes and added sorted keys again

agistrueai committed:
fixed imports

agistrueai committed:
removed extra lines from feeding_functions

agistrueai committed:
removed unnecessary import in feeding_functions

martinwicke commented on tensorflow/python/estimator/inputs/queues/feeding_functions.py:
```diff
@@ -21,6 +21,8 @@
 import random
 import numpy as np
+import six
+from types import FunctionType
```
Could you just import types and use types.FunctionType below?

agistrueai replied:
done. I use `import types as tp` because `types` is used as a variable inside the function

martinwicke commented on tensorflow/python/estimator/inputs/queues/feeding_functions.py:
```diff
@@ -235,8 +281,9 @@ def _enqueue_data(data,
   Args:
-    data: a numpy `ndarray`, `OrderedDict` of numpy arrays, or pandas
-      `DataFrame` that will be read into the queue.
+    data: a numpy `ndarray`, `OrderedDict` of numpy arrays,'Generator Function`
```
space after ','

agistrueai replied:
Done.

martinwicke commented on tensorflow/python/estimator/inputs/queues/feeding_functions.py:
```diff
@@ -218,6 +220,50 @@ def __call__(self):
+                                     list()).append(data_row[key])
+            list_dict_size += 1
+        feed_dict = {key: np.asarray(item)
+                        for key, item in list(list_dict.items())}
```
I would love some check that emits a very obvious error if the dicts emitted by the generator do not have identical key sets.

agistrueai replied:
Done.

martinwicke commented on tensorflow/contrib/learn/python/learn/learn_io/generator_io.py:
```diff
@@ -0,0 +1,132 @@
+    KeyError:  `target_key` not a key or `target_key[index]` not in next(`x()`).
+  """
+  if not isinstance(x, FunctionType):
+    raise TypeError('x must be generator function ; got {}'.format(
```
nit: no space before ';'

agistrueai replied:
Done everywhere as appropriate.

martinwicke commented on tensorflow/contrib/learn/python/learn/learn_io/generator_io.py:
```diff
@@ -0,0 +1,132 @@
+                  type(item).__name__))
+          if item not in input_keys:
+            raise KeyError(
+              'target_key or target_key[i] not in yielded dict ; got {}'.format(
```
I would omit 'or target_key[i]' here, and I would also output the keys present in the dict.

agistrueai replied:
Done

martinwicke commented on tensorflow/contrib/learn/python/learn/learn_io/generator_io.py:
```diff
@@ -0,0 +1,132 @@
+                  item))
+      else:
+        raise TypeError('target_key must be str or list of str ; got {}'.format(
+          type(target_key).__name__))
```
Do you have to use `__name__`? Does `str(type(target_key))` work?

agistrueai replied:
It's how it's being used everywhere else. e.g. inside feeding_functions
<img width="573" alt="screen shot 2017-03-29 at 15 57 00" src="https://cloud.githubusercontent.com/assets/23721105/24461246/73d144b8-1498-11e7-9ccb-e97e697d10a2.png">

martinwicke commented on tensorflow/contrib/learn/python/learn/learn_io/generator_io.py:
```diff
@@ -0,0 +1,132 @@
+  if target_key is not None:
+      if isinstance(target_key, str):
+        target_key = [target_key]
+      elif isinstance(target_key, list):
```
do you want to allow tuples and sets?

agistrueai replied:
Done. Not it  allows all containers

martinwicke reviewed (dismissed):
This is great. Final round of comments, I promise.

agistrueai committed:
fixed sorting of keys

added ability to pass any container of target_keys (e.g. list set tuple)
raise KeyError if mismatch between genarator dict keys
docstring fixes

agistrueai committed:
add empty line between class methods

agistrueai commented:
@martinwicke @ilblackdragon changes were done as requested please have a look at the review.

martinwicke committed:
Fixing some final formatting

agistrueai commented on tensorflow/python/estimator/inputs/queues/feeding_functions.py:
```diff
@@ -320,7 +320,7 @@ def _enqueue_data(data,
       ]
       queue_shapes = [()] + [col.shape[1:] for col in data.values()]
       get_feed_fn = _OrderedDictNumpyFeedFn
-    elif isinstance(data, tp.FunctionType):
```
if you look at lines 314,318 above types is being used as a variable and will overwrite the import that's why I imported it as tp

martinwicke committed:
Undo misguided types import

martinwicke commented:
Ah, the question about feeding_functions.py:323 is real: should it be FunctionType or GeneratorType?

agistrueai commented:
FunctionType that returns a GeneratorType. the feeder needs to be able to call it again if the generator gets consumed (for the next epoch)

martinwicke commented:
Jenkins, test this please?

martinwicke commented:
Wonderful. We will merge assuming the tests are happy. Thank you!

RealTimeDeployment commented:
I think I over did it a bit this morning on the security. The split was the better option. Need input functions for many different classes.

martinwicke commented:
@Realtimedeployment commenting for the wrong PR?

@agistrueai can you fix the indentation? See the sanity test failure for a full list.

RealTimeDeployment commented:
Sorry wrong pr.

agistrueai committed:
fixed indentation back to two spaces

agistrueai commented:
@martinwicke  sanity issues fixed. please retest

martinwicke commented:
Jenkins, test this please.

agistrueai commented:
@martinwicke  All tests passed. Just to reconfirm I verify consent to merge.

martinwicke commented:
Wonderful, thank you!
